boards: Add olimex_stm32_p103 board

Sturla22 opened the pull request:
Add support for the Olimex STM32 P103 board, which is the same PCB as P405 populated with an STM32F103 instead of an STM32F405.

I have tested blinky and hello_world on the board.

CAN support can most likely be added pretty easily.

erwango commented on boards/arm/olimex_stm32_p103/olimex_stm32_p103.yaml:
```diff
@@ -0,0 +1,13 @@
+flash: 128
+supported:
+  - counter
+  - can
```
Remove this if not supported in board dts

erwango commented on boards/arm/olimex_stm32_p103/pinmux.c:
```diff
@@ -0,0 +1,38 @@
+	{ STM32_PIN_PA2, STM32F1_PINMUX_FUNC_PA2_USART2_TX },
+	{ STM32_PIN_PA3, STM32F1_PINMUX_FUNC_PA3_USART2_RX },
+#endif
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(can1), okay)
+	{ STM32_PIN_PB8, STM32F1_PINMUX_FUNC_PB8_CAN1_RX },
+	{ STM32_PIN_PB9, STM32F1_PINMUX_FUNC_PB9_CAN1_TX },
+#endif
```
Remove this if not supported in board dts

erwango reviewed (changes requested):
Thanks for this contribution, 2 minors comments.

Also, please squash all you changes into one single commit, since PR are merged on top of master as is and  there is no interest to see all these steps in zephyr git tree.
Last, please add a commit message.

Sturla22 commented:
Thank you for reviewing @erwango, I'll fix what you pointed out. 
Should I remove the un-documented compatible strings pointed out by checkpatch as well? Or can I document them somewhere in the correct way? (I can't find anything similar in ./dts/bindings)

Sturla22 committed:
boards: Add olimex_stm32_p103 board

Add support for the Olimex STM32 P103 board, which is the same PCB as
P405 populated with an STM32F103 instead of an STM32F405.

Tested with blinky and hello_world on the board.

Signed-off-by: Sturla Lange <sturla22@gmail.com>

erwango commented on boards/arm/olimex_stm32_p103/Kconfig.board:
```diff
@@ -0,0 +1,9 @@
+# OLIMEX-STM32-P103 board configuration
+
+# Original work Copyright (c) 2017, Erwin Rol <erwin@erwinrol.com>
+# Modified work Copyright 2020, Sturla Lange
```
Please don't retain previous copyright for this kind of board files.
Indeed, what you pointed out as "original" was it self a copy of previous work ...
All these board files are the same and copied from each others.
I prefer to retain the copyright of the one who did the effort of porting and testing the board, instead of having a long list of copyrights...

erwango commented on boards/arm/olimex_stm32_p103/Kconfig.defconfig:
```diff
@@ -0,0 +1,12 @@
+# OLIMEX-STM32-P103 board configuration
+
+# Original work Copyright (c) 2017, Erwin Rol <erwin@erwinrol.com>
+# Modified work Copyright 2020, Sturla Lange
```
Ditto on copyright

erwango commented on boards/arm/olimex_stm32_p103/olimex_stm32_p103.dts:
```diff
@@ -0,0 +1,46 @@
+/*
+ * Original work Copyright (c) 2017, Erwin Rol <erwin@erwinrol.com>
+ * Modified work Copyright 2020, Sturla Lange
```
Ditto on copyright

erwango commented on boards/arm/olimex_stm32_p103/pinmux.c:
```diff
@@ -0,0 +1,34 @@
+/*
+ * Original work Copyright (c) 2017, Erwin Rol <erwin@erwinrol.com>
+ * Modified work Copyright 2020, Sturla Lange
```
Ditto on copyright

erwango reviewed (changes requested):
Sorry for this delay in reviewing again, github didn't notice that there were changes to review.
One minor comment about copyright.
Also, please add your name as CODEOWNER file in front of this board

martinjaeger commented on boards/arm/olimex_stm32_p103/pinmux.c:
```diff
@@ -0,0 +1,33 @@
+
+/* pin assignments for OLIMEX-STM32-P103 board */
+static const struct pin_config pinconf[] = {
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(usart2), okay)
```
```suggestion
#if DT_NODE_HAS_STATUS(DT_NODELABEL(usart2), okay) && CONFIG_SERIAL
```

martinjaeger commented on boards/arm/olimex_stm32_p103/olimex_stm32_p103.yaml:
```diff
@@ -0,0 +1,10 @@
+  - gnuarmemb
+  - xtools
+ram: 20
+flash: 128
```
```suggestion
flash: 128
supported:
  - gpio
```

martinjaeger commented on boards/arm/olimex_stm32_p103/olimex_stm32_p103_defconfig:
```diff
@@ -0,0 +1,34 @@
+
+CONFIG_SOC_SERIES_STM32F1X=y
+CONFIG_SOC_STM32F103XB=y
+# 64MHz system clock
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=64000000
```
This MCU runs on 72 MHz by default. Is there a reason to specify only 64 MHz for this board?

Please see also `nucleo_f103rb` board.

martinjaeger commented on boards/arm/olimex_stm32_p103/doc/index.rst:
```diff
@@ -0,0 +1,267 @@
+| GPIO      | on-chip    | gpio                 |
++-----------+------------+----------------------+
+
+Other hardware features are not supported by the Zephyr kernel.
```
That's not entirely true, as Zephyr should support quite a lot of this MCU. The features are just not configured in the devicetree of this board.

martinjaeger reviewed (changes requested):
Thanks for the addition of this board.

The failure of `tests/kernel/gen_isr_table` seems to be a general issue for this MCU. Please add the board to `platform_exclude` in `tests/kernel/gen_isr_table/testcase.yaml` to pass buildkite.

I suggest to add more features to the board.dts and pinmux, as otherwise the board is quite useless if only a single serial port and the LED is supported / enabled.

utzig commented on boards/arm/olimex_stm32_p103/olimex_stm32_p103_defconfig:
```diff
@@ -0,0 +1,34 @@
+CONFIG_CLOCK_STM32_PLL_MULTIPLIER=8
+CONFIG_CLOCK_STM32_AHB_PRESCALER=1
+CONFIG_CLOCK_STM32_APB1_PRESCALER=4
+CONFIG_CLOCK_STM32_APB2_PRESCALER=2
```
I use the following configuration to run everything at the highest allowed clock for this MCU:
```
CONFIG_CLOCK_STM32_PLL_MULTIPLIER=9
CONFIG_CLOCK_STM32_AHB_PRESCALER=1
CONFIG_CLOCK_STM32_APB1_PRESCALER=2
CONFIG_CLOCK_STM32_APB2_PRESCALER=1
```

martinjaeger replied:
Makes sense. These settings are the same as for [nucleo_f103rb](https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/nucleo_f103rb/nucleo_f103rb_defconfig).

martinjaeger commented on boards/arm/olimex_stm32_p103/olimex_stm32_p103.dts:
```diff
@@ -0,0 +1,45 @@
+
+/ {
+	model = "Olimex STM32-P103 board";
+
```
Please add also: 
```
	compatible = "olimex,stm32-p103", "st,stm32f103rb";
```
